File: ExodusImport/Source/ExodusImport/Private/JsonObjects/JsonRigidbody.cpp
#include "JsonImportPrivatePCH.h"
#include "JsonRigidbody.h"
#include "macros.h"

using namespace JsonObjects;

JsonRigidbody::JsonRigidbody(JsonObjPtr data){
	load(data);
}

void JsonRigidbody::load(JsonObjPtr data){
	/*I should probably get rid of those helper macros due to reduced error checking...*/
	JSON_GET_VAR(data, angularDrag);
	JSON_GET_VAR(data, angularVelocity);
	JSON_GET_VAR(data, centerOfMass);

	JSON_GET_VAR(data, collisionDetectionMode);
	JSON_GET_VAR(data, constraints);
	JSON_GET_VAR(data, detectCollisions);
	JSON_GET_VAR(data, drag);

	JSON_GET_VAR(data, freezeRotation);
	JSON_GET_VAR(data, inertiaTensor);
	JSON_GET_VAR(data, interpolation);

	JSON_GET_VAR(data, isKinematic);
	JSON_GET_VAR(data, mass);

	JSON_GET_VAR(data, maxAngularVelocity);
	JSON_GET_VAR(data, maxDepenetrationVelocity);

	JSON_GET_VAR(data, position);
	JSON_GET_VAR(data, rotation);

	JSON_GET_VAR(data, sleepThreshold);
	JSON_GET_VAR(data, solverIterations);
	JSON_GET_VAR(data, solverVelocityIterations);

	JSON_GET_VAR(data, useGravity);
	JSON_GET_VAR(data, velocity);
	JSON_GET_VAR(data, worldCenterOfMass);
}

File: ExodusExport/Scripts/Editor/Exporter/JsonRigidbody.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace SceneExport{
	[System.Serializable]
	public class JsonRigidbody: IFastJsonValue{
		public Rigidbody rig = null;
		public void writeRawJsonValue(FastJsonWriter writer){
			writer.beginRawObject();
			//The constant manual value dump is getting annoying. Need to think up a better way to do it.
			writer.writeKeyVal("instanceId", rig.GetInstanceID());

			writer.writeKeyVal("angularDrag", rig.angularDrag);
			writer.writeKeyVal("angularVelocity", rig.angularVelocity);
			writer.writeKeyVal("centerOfMass", rig.centerOfMass);

			writer.writeKeyVal("collisionDetectionMode", rig.collisionDetectionMode.ToString());
			writer.writeKeyVal("constraints", (int)rig.constraints);
			writer.writeKeyVal("detectCollisions", rig.detectCollisions);
			writer.writeKeyVal("drag", rig.drag);

			writer.writeKeyVal("freezeRotation", rig.freezeRotation);
			writer.writeKeyVal("inertiaTensor", rig.inertiaTensor);
			writer.writeKeyVal("interpolation", rig.interpolation.ToString());

			writer.writeKeyVal("isKinematic", rig.isKinematic);
			writer.writeKeyVal("mass", rig.mass);

			writer.writeKeyVal("maxAngularVelocity", rig.maxAngularVelocity);
			writer.writeKeyVal("maxDepenetrationVelocity", rig.maxDepenetrationVelocity);

			writer.writeKeyVal("position", rig.position);
			writer.writeKeyVal("rotation", rig.rotation);
			writer.writeKeyVal("sleepThreshold", rig.sleepThreshold);
			writer.writeKeyVal("solverIterations", rig.solverIterations);
			writer.writeKeyVal("solverVelocityIterations", rig.solverVelocityIterations);
			writer.writeKeyVal("useGravity", rig.useGravity);
			writer.writeKeyVal("velocity", rig.velocity);
			writer.writeKeyVal("worldCenterOfMass", rig.worldCenterOfMass);

			writer.endObject();
		}

		public JsonRigidbody(Rigidbody rig_){
			if (!rig_){
				throw new System.ArgumentNullException("rig_");
			}

			rig = rig_;
		}
	}
}
